test/extended/router/gatewayapicontroller.go

```diff
@@ -0,0 +1,491 @@
+
+		gwwAddress, err := oc.AsAdmin().WithoutNamespace().Run("get").Args("-n", "openshift-ingress", "gateway", gatewayName, "-o=jsonpath={.status.addresses[0].value}").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		e2e.Logf("The gateway Aaddress is: %v", gwwAddress)
+		o.Expect(lbExternalIP.Status.LoadBalancer.Ingress[0].Hostname).To(o.Equal(gwwAddress))
+
+		// get the dnsrecord name
+		dnsRecordName, err := oc.AsAdmin().WithoutNamespace().Run("get").Args("-n", "openshift-ingress", "dnsrecord", "-l", "gateway.networking.k8s.io/gateway-name="+gatewayName, "-o=jsonpath={.items[0].metadata.name}").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		e2e.Logf("The gateway API dnsrecord name is: %v", dnsRecordName)
+		// check whether status of published dns record of the gateway is True
+		dnsRecordstatus, err := oc.AsAdmin().WithoutNamespace().Run("get").Args("-n", "openshift-ingress", "dnsrecord", dnsRecordName, `-o=jsonpath={.status.zones[0].conditions[0].status}`).Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(dnsRecordstatus).To(o.Equal("True"))
+	})
+
+	g.It("Ensure HTTPRoute object is created", func() {
+		g.By("Getting the default domain for creating custom gateway")
+		defaultIngressDomain, err := getDefaultIngressClusterDomainName(oc, time.Minute)
+		o.Expect(err).NotTo(o.HaveOccurred(), "failed to find default domain name")
+		defaultDomain := strings.Split(defaultIngressDomain, "apps.")[1]
+
+		g.By("Confirm the default API Gateway is up")
+		_, gwerr := checkGatewayStatus(oc, gatewayName, "openshift-ingress")
+		o.Expect(gwerr).NotTo(o.HaveOccurred())
+
+		g.By("Create the http route using the default gateway")
+		defaultRoutename := "test-hostname.gwapi." + defaultDomain
+		createHttpRoute(oc, "test-httproute", defaultRoutename, "echo-pod-"+gatewayName)
+
+		g.By("Checking the http route using the default gateway is accepted")
+		assertHttpRouteSuccessful(oc, "test-httproute")
+	})
+})
+
+func checkGatewayClass(oc *exutil.CLI, name string) error {
+	gwapiClient := gatewayapiclientset.NewForConfigOrDie(oc.AdminConfig())
+
+	waitErr := wait.PollUntilContextTimeout(context.Background(), 2*time.Second, 2*time.Minute, false, func(context context.Context) (bool, error) {
+		gwc, err := gwapiClient.GatewayV1().GatewayClasses().Get(context, name, metav1.GetOptions{})
+		if err != nil {
+			e2e.Logf("failed to get gatewayclass %s, retrying...", name)
+			return false, nil
+		}
+		for _, condition := range gwc.Status.Conditions {
+			if condition.Type == string(gatewayapiv1.GatewayClassConditionStatusAccepted) {
+				if condition.Status == metav1.ConditionTrue {
+					return true, nil
+				}
+			}
+		}
+		e2e.Logf("Found gatewayclass %s but it is not accepted, retrying...", name)
+		return false, nil
+	})
+
+	if waitErr != nil {
+		return fmt.Errorf("Gatewayclass %s is not accepted", name)
+	}
+	e2e.Logf("Gateway Class %s is created and accepted", name)
```

Comment:
Also, why not use the `o.Expect(err).NotTo(o.HaveOccurred()` in these functions?
```suggestion
	o.Expect(waitErr).NotTo(o.HaveOccurred, "Gatewayclass %s is not accepted", name)	
```

Reply:
o.Expect(waitErr).NotTo(o.HaveOccurred(), "Gatewayclass %s is not accepted", name)